dl1_data_handler/reader.py

```diff
@@ -1,2028 +1,1206 @@
+                    if self.quality_query:
+                        passes_quality_checks = self.quality_query.get_table_mask(
+                            tel_table
+                        )
+                    # Merge the telescope table with the trigger table
+                    merged_table = join(
+                        left=tel_table[passes_quality_checks],
+                        right=trigger_table,
+                        keys=["obs_id", "event_id"],
                     )
-                if self.trigger_settings is not None:
-                    self.unprocessed_example_description.append(
-                        {
-                            "name": "trigger_patch_true_image_sum",
-                            "tel_type": self.tel_type,
-                            "base_name": "true_image_sum",
-                            "shape": (1,),
-                            "dtype": np.dtype(np.int),
-                        }
+                    table_per_type.append(merged_table)
+                table_per_type = vstack(table_per_type)
+
+                table_per_type = table_per_type.group_by(["obs_id", "event_id"])
+                table_per_type.keep_columns(self.example_ids_keep_columns)
+                if self.process_type == ProcessType.Simulation:
+                    tel_pointing = self._get_tel_pointing(f, self.tel_ids)
+                    table_per_type = join(
+                        left=table_per_type,
+                        right=tel_pointing,
+                        keys=["obs_id", "tel_id"],
                     )
+                    table_per_type = self._transform_to_spherical_offsets(table_per_type)
+                # Apply the multiplicity cut based on the telescope type
+                table_per_type = table_per_type.group_by(["obs_id", "event_id"])
 
-            if self.image_channels is not None:
-                self.unprocessed_example_description.append(
-                    {
-                        "name": "image",
-                        "tel_type": self.tel_type,
-                        "base_name": "image",
-                        "shape": self.image_mapper.image_shapes[
-                            self._get_camera_type(self.tel_type)
-                        ],
-                        "dtype": np.dtype(np.float32),
-                    }
-                )
-            if self.parameter_list is not None:
-                self.unprocessed_example_description.append(
-                    {
-                        "name": "parameters",
-                        "tel_type": self.tel_type,
-                        "base_name": "parameters",
-                        "shape": ((1,) + (len(self.parameter_list),)),
-                        "dtype": np.dtype(np.float32),
-                    }
-                )
+                def _multiplicity_cut_tel_type(table, key_colnames):
+                    self.min_telescopes_of_type.attach_subarray(self.subarray)
```

Comment:
bad design, you modify here some objects that are out of scope for this function. Also you don't use `key_colnames` local variable.

Reply:
`Astropy` filter functionality requires a function with those two arguments. See [here](https://docs.astropy.org/en/stable/table/operations.html#filtering).